src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationProvider.cs

```diff
@@ -721,26 +727,94 @@ private bool IsAuthenticationError(Exception ex)
             return false;
         }
 
-        private void UpdateCacheExpirationTime(KeyValueWatcher changeWatcher, bool success)
+        private void UpdateCacheExpirationTime(KeyValueWatcher changeWatcher)
         {
-            TimeSpan cacheExpirationTime;
+            TimeSpan cacheExpirationTime = changeWatcher.CacheExpirationInterval;
+            changeWatcher.CacheExpires = DateTimeOffset.UtcNow.Add(cacheExpirationTime);
+        }
 
-            if (success)
+        private async Task<T> ExecuteWithFailOverPolicyAsync<T>(Func<ConfigurationClient, Task<T>> funcToExecute, CancellationToken cancellationToken = default)
+        {
+            using IEnumerator<ConfigurationClient> clientEnumerator = _configClientManager.GetAvailableClients().GetEnumerator();
+
+            // No ConfigurationClients are available to execute the function.
+            if (!clientEnumerator.MoveNext())
             {
-                changeWatcher.RefreshAttempts = 0;
-                cacheExpirationTime = changeWatcher.CacheExpirationInterval;
+                return default;
             }

```

Comment:
I believe this was added because of the oddity of enforcing client provider to provide clients when all are backed off. I don't disagree with that sentiment. But for this helper method, I am concerned with breaching the contract outlined in `funcToExecute`. How can caller distinguish the case of funcToExecute returning default vs it coming from here?

Reply:
Spitballing here, if you pass clients you can avoid having to break the contract.

Reply:
Throwing an exception is still an option.

We use exceptions as a control flow mechanism in many places.
One common example is when any argument is null, we throw `ArgumentNullException`.
It is a similar thing here.
`ExecuteWithFailOverPolicy` needs a `configurationClient` to execute the `funcToExecute`, it throws an exception when there are no configuration clients available.

Reply:
As per our offline discussion, updated it to pass the enumerable of configuration clients to this method.